src/Swarm/Game/State.hs

```diff
@@ -260,6 +262,36 @@ instance Monoid (Notifications a) where
 
 makeLenses ''Notifications
 
+newtype Sha1 = Sha1 String
+
+data SolutionSource
+  = ScenarioSuggested
+  | -- | Includes the SHA1 of the program text
+    -- for the purpose of corroborating solutions
+    -- on a leaderboard.
+    PlayerAuthored Sha1
+
+data CodeToRun = CodeToRun SolutionSource ProcessedTerm
+
+getParsedInitialCode :: Maybe FilePath -> ExceptT Text IO (Maybe CodeToRun)
+getParsedInitialCode toRun = case toRun of
+  Nothing -> return Nothing
+  Just filepath -> do
+    contents <- liftIO $ TIO.readFile filepath
+    pt@(ProcessedTerm (Module (Syntax' srcLoc _ _) _) _ _) <-
+      ExceptT $
+        return $
+          left T.pack $
+            processTermEither contents
+    let strippedText = stripSrc srcLoc contents
+        programBytestring = TL.encodeUtf8 $ TL.fromStrict strippedText
+        sha1Hash = showDigest $ sha1 programBytestring
```

Comment:
If we are going to record hashes, I think we should hash some canonical representation of a program rather than the original source. 
* For example, at a basic level, we could first format the code by round-tripping through parsing + pretty-printing before hashing, or we could directly hash the AST.  That way, modifying the whitespace in your program would not change the hash.
* To make things slightly more interesting we could also do something like canonically rename all variables, so that renaming something in your program would not change the hash either.

However, I would be fine with leaving this to a future PR.